binding.cc

```diff
@@ -1,141 +1,211 @@
-  }
-
-  CRYPT_DATA_BLOB out = { 0, nullptr };
-  DWORD export_flags = EXPORT_PRIVATE_KEYS;
-  if (require_priv_key) {
-    export_flags |= REPORT_NO_PRIVATE_KEY | REPORT_NOT_ABLE_TO_EXPORT_PRIVATE_KEY;
-  }
-  if (!PFXExportCertStoreEx(memstore, &out, password, nullptr, export_flags)) {
-    ThrowWindowsError(env, "PFXExportCertStoreEx()");
-  }
-  Buffer<BYTE> outbuf = Buffer<BYTE>::New(env, out.cbData);
-  out.pbData = outbuf.Data();
-  if (!PFXExportCertStoreEx(memstore, &out, password, nullptr, export_flags)) {
-    ThrowWindowsError(env, "PFXExportCertStoreEx()");
-  }
-  
-  return outbuf;
-}
-
-// Export a given certificate from a system certificate store,
-// identified either by its thumbprint or its subject line.
-Value ExportCertificate(const CallbackInfo& args) {
-  std::vector<WCHAR> password_buf = MultiByteToWideChar(args[0].ToString());
-  LPCWSTR password = password_buf.data();
-  std::string sys_store_name = args[1].ToString();
-  HCERTSTORE sys_cs = CertOpenSystemStoreA(0, sys_store_name.c_str());
-  if (!sys_cs) {
-    ThrowWindowsError(args.Env(), "CertOpenSystemStoreA()");
-  }
-  Cleanup cleanup_sys_cs([&]() { CertCloseStore(sys_cs, 0); });
-  
-  PCCERT_CONTEXT cert = nullptr;
-  Object search_spec = args[2].ToObject();
-  if (search_spec.HasOwnProperty("thumbprint")) {
-    Buffer<BYTE> thumbprint = search_spec.Get("thumbprint").As<Buffer<BYTE>>();
-    CRYPT_HASH_BLOB thumbprint_blob = { 
-      static_cast<DWORD>(thumbprint.Length()),
-      thumbprint.Data() 
-    };
-    cert = CertFindCertificateInStore(
-        sys_cs,
-        X509_ASN_ENCODING | PKCS_7_ASN_ENCODING,
-        0,
-        CERT_FIND_HASH,
-        &thumbprint_blob,
-        nullptr);
-  } else if (search_spec.HasOwnProperty("subject")) {
-    cert = CertFindCertificateInStore(
-        sys_cs,
-        X509_ASN_ENCODING | PKCS_7_ASN_ENCODING,
-        0,
-        CERT_FIND_SUBJECT_STR,
-        MultiByteToWideChar(search_spec.Get("subject").ToString()).data(),
-        nullptr);
-  } else {
-    SetLastError(CRYPT_E_NOT_FOUND);
-  }
-  if (!cert) {
-    ThrowWindowsError(args.Env(), "CertFindCertificateInStore()");
-  }
-
-  Cleanup cleanup_cert([&]() { CertFreeCertificateContext(cert); });
-
-  return CertToBuffer(args.Env(), cert, password, args[3].ToBoolean());
-}
-
-}
-
-static Object InitWinExportCertAndKey(Env env, Object exports) {
-  exports["exportCertificate"] = Function::New(env, ExportCertificate);
-  return exports;
-}
-
+#include <napi.h>
+#include <windows.h>
+#include <wincrypt.h>
+
+using namespace Napi;
+
+namespace {
+
+// Naive RAII cleanup helper.
+struct Cleanup {
+  std::function<void()> fn;
+  Cleanup(std::function<void()> fn) : fn(fn) {}
+  ~Cleanup() { fn(); }
+};
+
+// Convert UTF-8 to a Windows UTF-16 wstring.
+std::wstring MultiByteToWideChar(Value value) {
+  static_assert(sizeof(std::wstring::value_type) == sizeof(std::u16string::value_type),
+      "wstring and u16string have the same value type on Windows");
```

Comment:
Yes, exactly. I don’t think this is a breaking change that could realistically ever happen, so this is mostly just very “active” documentation about the fact that this equality holds on Windows (which may not be obvious to developers who are not used to Windows, because on other platforms “wide chars” are 32-bit, not 16-bit).

Reply:
Gotcha! Yeah, makes sense, I like this pattern of "active" documentation 😃